week08_pomdp/practice_pytorch.ipynb

```diff
@@ -46,28 +65,32 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
-   "outputs": [
-    {
-     "name": "stdout",
-     "output_type": "stream",
-     "text": [
-      "\u001b[33mWARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.\u001b[0m\n",
-      "Observation shape: (1, 42, 42)\n",
-      "Num actions: 14\n",
-      "Action names: ['NOOP', 'UP', 'RIGHT', 'LEFT', 'DOWN', 'DOWNRIGHT', 'DOWNLEFT', 'RIGHTFIRE', 'LEFTFIRE', 'DOWNFIRE', 'UPRIGHTFIRE', 'UPLEFTFIRE', 'DOWNRIGHTFIRE', 'DOWNLEFTFIRE']\n"
-     ]
-    }
-   ],
+   "outputs": [],
    "source": [
+    "import time\n",
+    "import warnings\n",
+    "from IPython.display import clear_output\n",
+    "from collections import defaultdict\n",
+    "from typing import Text, List, Dict, Set, Union\n",
+    "\n",
     "import gym\n",
+    "from gym.wrappers.monitoring.video_recorder import VideoRecorder\n",
+    "from gym.wrappers import RecordVideo, RecordEpisodeStatistics\n",
+    "\n",
     "from atari_util import PreprocessAtari\n",
+    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
     "\n",
+    "# We scale rewards to avoid exploding gradients during optimization. [ISSUE No.04 FACED ME]\n",
```

Comment:
We can't merge into `master` something that says "faced me". Please clarify what exactly is wrong here, why this scaling is needed (I myself vaguely remember that something like this was here, but I don't remember the details at all), and amend the comment.

Reply:
Tbh, it was one of the **PROBLEMS** that took me two months to recognize. Without the **Reward Scaling**, the gradient at some step will start to magnify due to the scale of the rewards when it starts to increase; this will make the network start the path of **Gradient Vanishing**, and the network will begin to stop learning. Therefore, the Agent will stop learning anymore.

In this environment, I notice that reducing the magnification of the **Gradient normalization** helps to learn faster and more robustly; that's why; you can find me adding **Layer Normalizationation** in the Agent's network, too. I also tried adding a **value coefficient** for a stable learning path. Honestly, **Policy Gradient** will be an experience that I will never forget in my whole life!

I have read books and papers for four months to solve `week_06]` and `week_08`. I am proud of that.